nn/transformer.py

```diff
@@ -0,0 +1,286 @@
+  """
+  def __init__(self, output_dim: int = 512, num_heads: int = 8, num_encoder_layers: int = 6,
+               num_decoder_layers: int = 6, dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu,
+               custom_encoder: Optional[Any] = None, custom_decoder: Optional[Any] = None,
+               custom_encoder_layer: Optional[nn.LayerRef] = None, custom_decoder_layer: Optional[nn.LayerRef] = None,
+               norm_eps: float = 1e-5, norm=nn.layer_norm, encoder_pos_enc: Callable = nn.positional_encoding,
+               decoder_pos_enc: Callable = nn.positional_encoding) -> None:
+    """
+    :param output_dim: output dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param num_encoder_layers: Number of encoder layers
+    :param num_decoder_layers: Number of decoder layers
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param custom_encoder: Custom Encoder to replace the standard encoder
+    :param custom_decoder: Custom Decoder to replace the standard decoder
+    :param custom_encoder_layer: Custom Encoder layer to replace the standard layer if custom_encoder and
+      custom_encoder_layer are given custom_encoder will be preferred
+    :param custom_decoder_layer: Custom Decoder layer to replace the standard layer if custom_decoder and
+      custom_decoder_layer are given custom_decoder will be preferred
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm: function for layer normalization
+    :param encoder_pos_enc: Positional Encoding for the Encoder
+    :param decoder_pos_enc: Positional Encoding for the Decoder
+    :param att_dropout: dropout value for attention
+    """
+    super().__init__()
+
+    if custom_encoder is not None:
+      self.encoder = custom_encoder
+    else:
+      if custom_encoder_layer is not None:
+        encoder_layer = custom_encoder_layer
+      else:
+        encoder_layer = TransformerEncoderLayer(
+          dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+          norm_eps=norm_eps, norm=norm, att_dropout=att_dropout)
+      self.encoder = TransformerEncoder(
+        encoder_layer=encoder_layer, num_layers=num_encoder_layers, norm=norm, norm_eps=norm_eps,
+        pos_enc=encoder_pos_enc)
+
+    if custom_decoder is not None:
+      self.decoder = custom_decoder
+    else:
+      if custom_decoder_layer is not None:
+        decoder_layer = custom_decoder_layer
+      else:
+        decoder_layer = TransformerDecoderLayer(
+          dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+          norm_eps=norm_eps, norm=norm, att_dropout=att_dropout)
+      self.decoder = TransformerDecoder(
```

Comment:
I wonder now what we should use here.

We can either define the whole `Transformer` step-wise (as `TransformerStep` or so) and then use `TransformerDecoderStep` here. But I wonder if this is useful.

Or we also define a `TransformerDecoder`. But what should it do? Use a ground truth input? But this only makes sense in training and not in recognition.

I think the `Transformer` should be useful for both cases. So there might be some `search` option to the module which defines whether it uses a ground truth target or uses search. Or some other way to determine this (see #18).

We should maybe abstract that away, i.e. a generic interface from the decoder output to some `choice` module in search, and using the prev output of that as input to an embedding, and in training using the ground truth as input. Although that should be a bit more generic, also covering the regression case (TTS).

Reply:
How would you initialized the state for the TransformerDecoderStep? 
So was thinking in Transformer something like: 
```
  def forward(self, source: nn.LayerRef, target: nn.LayerRef) -> nn.LayerRef:
    """
    Forward step of Transformer
    """
    memory = self.encoder(source)
    with nn.Loop() as loop:
      target_ = loop.unstack(target, axis="T")
      memory_ = loop.unstack(memory, axis="T")
      loop.state.decoder = nn.State(initial=0)
      output_, state = self.decoder(target_, memory=memory_, state=loop.state.decoder)
      output = loop.stack(output_)

    return output
```
But I am not sure about the `loop.state.decoder` line, do we have to initalize each of the recurrent parameters by hand? so in this case `k_accum` and `v_accum` or also others I am missing right now? Is the Stacking/Unstacking even correct?

Reply:
`memory` has a different axis, and you would not unstack that put pass it as a whole.

`target` is only available during training. Or more specific, it must be provided when you don't use search, but with search it would use the previous predicted. How we define `search` is also still open (#18). Maybe just a `search: bool` option to the module.

For recurrent modules, we have the convention that they should provide an `initial_state` function. So you would do:
```
loop.state.decoder = nn.State(initial=self.decoder_initial_state())
```

The return value of `self.decoder(...)` would also be assigned again to `loop.state.decoder`.

The target labels also need some state because you must access the previous value.
Also you would return the last state. And it should accept the initial state.
So:
```
def forward(..., initial_state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
  ...
    loop.state = initial_state
    logits, loop.state.decoder = self.decoder(loop.state.target, memory=memory, state=loop.state.decoder)
    if self.search:
      loop.state.target = nn.choice(logits, input_type="logits")
    else:
      loop.state.target = loop.unstack(target)
    outputs = loop.stack(loop.state.target)
  return outputs, loop.state
```
(Just a draft. I think there are multiple things here which still would not work correctly like that.)

Reply:
And the `initial_state` of `Transformer` would look sth like this:
```
def initial_state(initial_target: nn.LayerRef = 0) -> nn.LayerState:
  return nn.LayerState(
    target=initial_target,
    decoder=self.decoder.initial_state())
```

And `initial_state` of the `TransformerDecoderStep` would collect the initial state of all the layers.

And `initial_state` of each `TransformerDecoderLayerStep` would use the `initial_state` of the `SelfAttentionStep`.
(Which is also missing, but this will be implemented.)

Reply:
I think I get what you mean. I will code a draft and push it, so we can then from there further refine it.